repair/repair.cc

```diff
@@ -1993,23 +2037,94 @@ future<> repair_service::do_rebuild_replace_with_repair(locator::token_metadata_
             }
             auto& strat = erm->get_replication_strategy();
             dht::token_range_vector ranges = strat.get_ranges(myid, *tmptr).get();
-            auto& topology = erm->get_token_metadata().get_topology();
             std::unordered_map<dht::token_range, repair_neighbors> range_sources;
             auto nr_tables = get_nr_tables(db, keyspace_name);
-            rlogger.info("{}: started with keyspace={}, source_dc={}, nr_ranges={}, ignore_nodes={}", op, keyspace_name, source_dc, ranges.size() * nr_tables, ignore_nodes);
+            sstring source_dc_for_keyspace;
+            // Allow repairing in the source_dc only if there are enough replicas remaining
+            if (!source_dc_live_nodes.empty()) {
+                switch (strat.get_type()) {
+                case locator::replication_strategy_type::network_topology: {
+                    const auto& nt_strat = dynamic_cast<const locator::network_topology_strategy&>(strat);
+                    size_t rf = nt_strat.get_replication_factor(source_dc->name);
+                    size_t nr_racks = source_dc_racks.size();
+                    // max_replicas_per_rack calculation is based on NetworkTopologyStrategy
+                    // replica selection algorithm which fills rack in round robin
+                    // so the maximum replicas per rack is ceil(rf / nr_rack).
+                    size_t max_replicas_per_rack = (rf + nr_racks - 1) / nr_racks;
+                    size_t possibly_lost_replicas = 0;
+                    for (const auto& [rack, nr] : lost_nodes_per_rack) {
+                        possibly_lost_replicas += std::min(nr, max_replicas_per_rack);
+                    }
+                    // Can we ensure a quorum in source_dc?
+                    if (possibly_lost_replicas < rf / 2) {
+                        source_dc_for_keyspace = source_dc->name;
+                    } else if (reason == streaming::stream_reason::rebuild) {
+                        if (rf == 0) {
+                            auto error_msg = format("Cannot rebuild keyspace={} using source_dc={}: Replication factor in this datacenter is 0",
+                                    keyspace_name, source_dc_desc);
+                            if (source_dc->user_provided) {
+                                throw std::runtime_error(format("{}. {}", error_msg, reselect_dc_msg));
+                            } else {
+                                rlogger.warn("{}. {}", error_msg, fallback_msg);
+                                source_dc.reset();
+                            }
+                        } else {
+                            auto error_msg = format("It is unsafe to rebuild keyspace={} using source_dc={}: Possibly lost {} replicas out of {}.",
+                                    keyspace_name, source_dc_desc, possibly_lost_replicas, rf);
+                            if (possibly_lost_replicas >= rf) {
```

Comment:
Why do we check possibly_lost_replicas >= rf ?

Reply:
Since we suspect there are no more replicas left and that the user may have made a mistake (hence the throw below).
If there are remaining replicas, we warn the user below and use their choice anyway.

Reply:
Say we have 5 racks, each have 2 nodes. With RF = 3, we have 3 nodes down first 3 racks. In this case, possibly_lost_replicas = 3. However, it is quite possible, the down nodes are not responsible for any data and the rebuild could performed without any issue. However,  the check below reject it.

Reply:
Yes, the checks are for the worst case since it is keyspace-wide. Both with vnodes and with tablets the replica-set is per-token so we can't assume any locality of particular replicas on particular racks or nodes.